### ISSUE TITLE
[NPC] Allow allied NPC's to be tasked to complete nearby "craft in progress" items

### ISSUE THREAD

**Issue by Zourin2** [CleverRaven/Cataclysm-DDA#49472]
Currently, NPC's are mostly only capable of performing craft work via the Base Camp system, or with the direct and active intervention of the player (craft assist/watching over the shoulder).  NPC's can be tasked with construction work, lumber tasks, and even reading skillbooks, but incapable of being told "Finish boiling this water while I'm gone."

# Describe the solution you'd like
A new task option similar to construction or book reading that allows a player to request an allied NPC to finish a 'craft in progress' item (initiated by the player), provided it falls within their skill limits. For example, being asked to complete a 50-unit batch craft of Clean Water or Salt. I would expect these activities to be interrupted much like the reading task would be, based on sleep needs or nearby enemies.

This solution allows the simple, straightforward assignment of long, simple jobs to NPC's within their capabilities while the player engages more active content. Realistically, this gives Specialist NPC's a more active, valuable, and autonomous role in their mutual-interest survival with the player without needing a 'base camp city' built first, and an early game avenue of developing practical non-combat skills in non-specialist NPC's.

# Describe alternatives you've considered
Alternatives are in place, but are locked behind the dramatically costly base camp systems or require being babysat by the player. Base camps can/do offer a different array of activities and quality of life options and are better for managing the operations of a larger number of allied NPC's. 

I've noted several closed/failed attempts at expanding NPC activity, so hopefully this may see some traction.

**Comment by natsirt721** [comment]
Big fan of this idea, but there are a couple of corner cases here to consider:

- The NPC will have to check to see if they have the correct minimum skills before beginning the craft (well, resuming it but beginning for them). This should be pretty easy though.
- What does partial failure look like? If they fail once will they abort? Will they keep eating materials until they succeed or none are left?
- What about liquids? What if there is no container nearby to put the liquid in? Or multiple containers? Or containers that are partially full of the same liquid?
- What about tools? Can the player/NPCs use the same tool to work on the different crafts simultaneously? What about tools with charges? Can NPCs keep a nearby fire going to keep cooking/boiling?

**Comment by Zourin2** [comment]
- NPC's should check for tool availability, which is also checked when you create the 'craft in progress' item. players are responsible for leaving tools available for any jobs they want the NPC to handle.  
- NPC's shouldn't do jobs that have a chance of failure, so are generally limited to jobs of their level or lower. They can't 'craft up' like the player, even if they could assist (which I'm not sure if they can).
- NPC's should prefer placing liquid results in existing partially filled containers before using the nearest, largest container(s) available. remainder can be dumped. Again, player responsibility for making sure the NPC has all the tools and means available.
- NPC's should be able to share tools that are within the usual crafting radius. They should also be able to utilize firewood/fuel zones to maintain fires, although the player should at least get one lit for them. 
- Player is responsible for placing the 'craft in progress' item on a table, folding table, crafting spot, or workbench if they want those bonuses utilized, if possible. Kudo bonus for a 'crafting zone' similar to a vehicle deconstruction zone or woodcutting zone.

**Comment by natsirt721** [comment]
> * NPC's should check for tool availability, which is also checked when you create the 'craft in progress' item. players are responsible for leaving tools available for any jobs they want the NPC to handle.

Tool availability isn't tracked per-craft, its binary. Either you have the tool or you don't. With no changes, 100 NPCs crafting 100 items could share one hammer.

> * NPC's shouldn't do jobs that have a chance of failure, so are generally limited to jobs of their level or lower. They can't 'craft up' like the player, even if they could assist (which I'm not sure if they can).

With the recent-ish changes to the crafting system and the introduction of proficiencies, most recipes have a small-but-nonzero chance of failure even for 'qualified' crafters. Determining at what proficiency availability/skill level that can be mitigated is nontrivial. In any event, this will require NPCs to be overqualified (over-leveled and probably with one or more contributing proficiencies) in order to craft. When the crafting system is eventually migrated to use `normal_roll` this we be even less easy to gauge. 

> * NPC's should prefer placing liquid results in existing partially filled containers before using the nearest, largest container(s) available. remainder can be dumped. Again, player responsibility for making sure the NPC has all the tools and means available.

Makes sense. The dump-extra is harsh but there really is no other way around it.

> * NPC's should be able to share tools that are within the usual crafting radius. They should also be able to utilize firewood/fuel zones to maintain fires, although the player should at least get one lit for them.

This would require an otherwise separate enhancement in-and-of itself (NPCs keeping fires alive). Might lead to weird interactions or excessive fuel use with multiple NPCs + Player trying to feed the same fire, but that can probably be mitigated. Agree that the player should have to light the fire initially.

**Comment by PatrikLundell** [comment]
Another factor that will have to be considered is how to proceed when the companion is outside of the reality bubble. The base camp crafting system dodges that by having the companion disappear and be called back actively by the player after the crafting time has expired. Allowing for companion crafting outside of the base camp work around would probably require actually solving the problem rather than dodging it (deconstruction zones seem to have activities suspended while the player is away and then catch up when the companions come within the reality bubble, but with crafting you have the additional issue of ensuring the crafting resources are within the bubble as well so they can be deducted properly. Using the base camp bulletin board ensures the player is located within two tiles of the location the player was at when the order was given, reducing the risk of resource availability edge cases [but I wouldn't be surprised if you'd still run into such cases if you placed your zones and items carefully to have resources available when ordering crafting but not when recalling the companion, by standing on opposite sides of the board]).

**Comment by Zourin2** [comment]
I'd say the reality-bubble issue is tolerable. The idea is mostly to get the NPC's to operate more in parallel with the player survivor.

Infinite scaling Tool-sharing is likewise tolerable, as most tools of some concern involve fuel, such as the soldering iron and welding station (power reserve). Once the fuel runs out, the craft halts at its current progress. The idea is mostly to revolve around a small number of specialist/trained NPC's to be performing different functions. It's the players responsibility to provide resources to scale with their team.

Fire-fuel management already exists as a part of player activity, so in some effect, the code is already written in part for it to happen. This may be something where we 'wait and see' how multiple NPC's handle it. It may be as simple as the npc nearest the firewood source declaring 'i got it' and halting any further effort.

Small-but-nonzero chance of failure is acceptable. The limit is mostly on the ability to craft slightly above your level with a substantially higher chance of failure, and as a stepping stone for further features like Tailor specialists going around and repairing sorted, clean clothing (which has infamous failure rates) or Medic specialists patching up injured allies.

**Comment by Zourin2** [comment]
Given the scale of the request, I've added a bounty for this issue if seen fit to take up.

**Comment by PatrikLundell** [comment]
My parallel usage is to have companions stay in the base camp and do what they can do, while my character collects resources. I gave up trying to use companions for assistance on trips when I realized they won't run after the PC when retreating (just walk), and there's a lot of backtracking in my zombie fighting...
Requiring PC presence isn't much of an issue when actually doing parallel crafting, but setting them off to do a big batch of some kind of slow crafting is. I'm not saying it's useless, but there are significant restrictions that ought to be made clear to the player. You probably wouldn't want them to start of a multi day batch crafting without issuing a warning, for instance.

When it comes to failures, I think the companion order has to include a selection concerning failures, e.g. don't pause until out of resources/pause when less than X resources left, pause if X failed attempts, pause on first failure.

**Comment by Zourin2** [comment]
@PatrikLundell NPC's can't do crafting without a base camp or assisting/observing the player. The biggest effect of enabling them to finish initiated jobs is that you can have them handle long jobs like making clean water, salt, nut butter, sugar, slingshot ammo like marbles or clay pellets, or even stupid long jobs like large wind turbines. Generally, if you have an NPC with sufficient skill, you can delegate these 'day burners' if you have other priorities they can't handle, like scavenging, foraging, or hunting.

They can already be assigned super-long tasks like vehicle dismantling, woodcutting (which usually results in more chips than wood), or even reading a book. In the latter case, they'll happily interrupt their job for a nap and have to be reassigned, which is about the level of PC management i'm aiming for. The important thing is that they're more functional as 'leadership-craving survivors' and a little less "meat furniture".

**Comment by chrispikula** [comment]
A few more corner cases:
If materials change during crafting, how to handle?  Eg: food rotting, or other more esoteric tick-based transforms.  I'm guessing the answer is to stop the recipe, but it should be handled gracefully, and non-food transforms might need a custom check of some sort.

What happens if the tools disappear during crafting as well?  If I tell someone to make 50 trenchcoats, they're going to be at it for a month.  If I decided, meanwhile, that I'm going to take their tools, how is that to be handled?  If they were available to be shared, that is, not merged into the task, what's stopping me from just taking them?

Regarding charged tools above and my second point, how would fallback tools be handled?  If there's 4 soldering irons, each with 75 charges, and the job needs 150, do two get used up?  If one gets taken, then moves onto a third?  Two hammers, and the one being used gets 'taken'?
A downside of searching the entire local area for tools, (and updating their charges) each 5-minute crafting tick would result in significant lag spikes.  Especially in highly cluttered areas, like what my crafting zone looks like.  :-p  With 4 or 5 crafters all going at it, this would be harsh.

I honestly think that NPC crafting should temporarily consume tools during the craft, so as to not deal with the above.  PC, it can makes sense that it doesn't, as we can stop crafting, then get back to it, and are even expected to do so.  Two NPC's working simultaneously on separately crafting, say, anvil requiring recipes?  How are they sharing it?  Or, say, when making clothes, a sewing kit.  Do they pass the needle back and forth?

It's not like it's hard to make duplicates of the tools, and it makes sense.

**Comment by PatrikLundell** [comment]
Some comment on @chrispikula's comment:

Food rotting: All ingredients are allocated and, I think, converted into the crafting in progress item, although I haven't examined food in progress items. Regardless, I don't think it's an issue exclusive to companion crafting, as you could run out of fire half way through and try to resume your foie gras recipe a month later (and I don't know what happens).

Ideally tools should be allocated to the project and become unavailable for anything else while crafting is active, but be released when crafting is paused. Resuming should require the appropriate tools again (if it's a multi stage project you really should need only the ones required for the current stage). Again, it's an issue common to base camp and PC crafting as well. I'd expect current crafting resumption to require the tools again, so if you took the McGuffin halfway through the companion shouldn't be able to resume the crafting. Base camp crafting dodges the issue by checking at the start and have the companion disappear for the crafting duration.

The charged resources issue is common to base camp crafting as well (again, that process cheats), and is really part of the "lock resourced into active crafting" issue. I'd aim for allocating tools as well as charge containers required for the remainder of the project to the project and release the tools and more or less depleted charge containers when the crafting job is paused.
Technically, I think that would result in crafting creating "active in progress" items that are then turned into "in progress" items when paused (actively by the PC, passively because it's time to eat/sleep/etc. by the companion). Resuming would again locate and lock up resources in the object that turns back into an active in progress one. Locked up resources should probably be marked as "in use" and shouldn't be possible to pick up or move, but you'd still need to handle destruction (a gasoline zombie blowing up your fuel stores should result in your fuel dependent crafting being suspended). It can be noted that by locking up resources at the start/resumption of crafting, you're not really dependent on the reality bubble, but could just fast-forward the companion's crafting progress when you return. The issue that doesn't cover is failures, however, as that would either require allocation of failure buffer resources when resuming the crafting, or allocating additional resources when failing (which can be done during fast-forwarding, assuming the resources are within the reality bubble together with the companion).

It seems we're thinking along similar lines here.

**Comment by chrispikula** [comment]
Food can rot and make the result rotten.  I'm not sure if this is a special case of transforming items or not.  I can't think of any items that you can use in a crafting recipe during their transformation step, eg, fermenting vinegar or yeast starters.  

I'm thinking that there needs to be a graceful method in play for ingredients becoming invalid, otherwise it'd be hell to try to track down that bug later, doubly so if it's got reality-bubble & fast-forwarding complications going on.

**Comment by Zourin2** [comment]
iirc, 
> All ingredients are allocated and, I think, converted into the crafting in progress item
This, and I'm not 100% sure if charges are consumed when the 'craft in progress' item is consumed as well.

as far as I'm concerned, most 'edge cases' can simply end in the halting of progress on the item. When the player attempts to re-initiate the craft, the NPC can explain he's missing charges, tools, etc when it halts again.

**Comment by adamaxis** [comment]
I'm working on a similar system, but which utilizes the base camp calorie system (crafting+assembly construction). NPCs can be told to craft or assemble something without it, but they get angry based on how many calories it cost them(it uses the existing base camp calorie consumption).

**Comment by adamaxis** [comment]
> Currently, NPC's are mostly only capable of performing craft work via the Base Camp system, or with the direct and active intervention of the player (craft assist/watching over the shoulder). NPC's can be tasked with construction work, lumber tasks, and even reading skillbooks, but incapable of being told "Finish boiling this water while I'm gone."
> 
> # Describe the solution you'd like
> A new task option similar to construction or book reading that allows a player to request an allied NPC to finish a 'craft in progress' item (initiated by the player), provided it falls within their skill limits. For example, being asked to complete a 50-unit batch craft of Clean Water or Salt. I would expect these activities to be interrupted much like the reading task would be, based on sleep needs or nearby enemies.
> 
> This solution allows the simple, straightforward assignment of long, simple jobs to NPC's within their capabilities while the player engages more active content. Realistically, this gives Specialist NPC's a more active, valuable, and autonomous role in their mutual-interest survival with the player without needing a 'base camp city' built first, and an early game avenue of developing practical non-combat skills in non-specialist NPC's.
> 
> # Describe alternatives you've considered
> Alternatives are in place, but are locked behind the dramatically costly base camp systems or require being babysat by the player. Base camps can/do offer a different array of activities and quality of life options and are better for managing the operations of a larger number of allied NPC's.
> 
> I've noted several closed/failed attempts at expanding NPC activity, so hopefully this may see some traction.

Are you looking for a simple delegation system? The solution I'm working on is basically a simple delegation system that does a few different things, some outside of the scope of this request:
1) All crafting actions (outside of batch but including assemblies) all use the same expense formula: calorie cost(2500/24/60/60) + (100/5). I didn't want to bloat out the craft function anymore than it already is, so I tied costs into the existing five_step system.
2) Handles NPC crafting with the same menu and system that the player uses. Thus, this system utilizes skill and skill-books in the same manner as players.
3) Handles liquids and liquid crafting in a prefetch manner using the handle_liquid system with some slight modifications, with the default behavior (in the event a prefetched container can't be found) being 'stick them wherever they'll fit' and secondarily 'dump it on the ground'.
4) Handles assembly crafting using the same system as players. Skill and skill-books are factored into this.

**Comment by Zourin2** [comment]
I'm mostly looking for a simple task delegation (ala 'read this book') outside the basecamp system. Basecamps are still in a 'developmental' state and not something you should be particularly diving into with only 1-2 followers. Having 1 follower is nearly useless unless they're a godsend at combat.  

Being able to ask a lone follower to complete a prepared 'work in progress' (created by the player) for which they are qualified vastly improves their value and utility, even if they are sensitive to failure states such as a lack of fuel, light, or exhaustion to start or continue the task, not dissimilar their book-reading behavior.

**Comment by adamaxis** [comment]
> I'm mostly looking for a simple task delegation (ala 'read this book') outside the basecamp system. Basecamps are still in a 'developmental' state and not something you should be particularly diving into with only 1-2 followers. Having 1 follower is nearly useless unless they're a godsend at combat. basecamps aren't an option if, for example, your personal 'home camp' is a basement or a basecamp-incompatible building.

I feel like we're on the same page with this. My system is basically complete.
 
> Being able to ask a lone follower to complete a prepared 'work in progress' (created by the player) for which they are qualified vastly improves their value and utility, even if they are sensitive to failure states such as a lack of fuel, light, or exhaustion to start or continue the task, not dissimilar their book-reading behavior.

The issue I have with delegation w/o calorie bartering is that you essentially become a dictator. I don't think many people would work for free in this way, especially not in Cataclysm's hectic-and-often-bipolar world of personal relations. There are obvious benefits to being under the protection of a group, but we don't have a system to put a value to these kind of exchanges. I don't currently see a better medium of exchange to calories - value doesn't currently work because nothing is expended in the process and you can just take everything back after you barter it. I personally think a more sophisticated relations system to encompass the transactional nature of relationships in the post-apocalypse would be lovely.

Do you have some alternative to calorie bartering? May be able to utilize the food aspect of basecamp without constructing one.

> Basecamps do allow for a level of remote management and much more reliable self-automation, so its value isn't completely moot even with this feature. Long term, Delegation and Basecamps would compliment and scale quite well in combination with a team of specialist npc followers.

100%

**Comment by Standing-Storm** [comment]
Implemented by #67458

**Comment by hungpham2411** [comment]
should this be closed now?
